File: Presentation/ClientApp/src/app/Models/InvoiceDetail.ts
import { Product } from './product.model';

export class InvoiceDetail {
    idDetail: number;
    unitValue: number;
    quantity: number;
    discount: number;
    tolalDetail: number;
    idInvoice: string;
    idProduct: string;
    product: Product;

    constructor (product: Product, quantity: number, discount: number, price: number) {
        this.product = product;
        this.unitValue = price;
        this.idProduct = product.idProduct;
        this.quantity = quantity;
        this.discount = discount;
        this.idDetail = 0;
        this.CalculateTotalDetail();
    }

    CalculateTotalDetail (): void {
        this.tolalDetail = (this.quantity * this.product.salePrice) * (1 - (this.discount / 100));
    }

    CalculateIva(): number {
        return this.tolalDetail * (this.product.iva / 100);
    }

}

File: Entity/InvoiceDetail.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Entity
{
    public partial class InvoiceDetail
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int IdDetail { get; set; }
        [Column(TypeName = "decimal(18,2)")]
        public decimal UnitValue { get; set; }
        public float Quantity { get; set; }
        public float Discount { get; set; }
        [Column(TypeName = "decimal(18,2)")]
        public decimal TolalDetail { get; set; }
        [NotMapped]
        public string IdInvoice { get; set; }
        [NotMapped]
        public string IdProduct { get; set; }
        [Column(TypeName = "nvarchar(10)")]
        [ForeignKey("IdProduct")]
        public virtual Product Product { get; set; }
        
        public InvoiceDetail()
        {
            
        }

        public InvoiceDetail(Product product, float quantity, float discount, decimal price)
        {
            IdProduct = product.IdProduct;
            Product = product;
            UnitValue = price;
            Quantity = quantity;
            Discount = discount;
            CalculateTotalDetail();
            CalculateIva();
        }

        public void CalculateTotalDetail()
        {
            
            TolalDetail = Decimal.Round((((decimal)Quantity) * Product.SalePrice) * (1 - ((decimal)Discount/100)), 1);
        }


        public decimal CalculateIva()
        {
            return TolalDetail * (Product.Iva  / 100) ;
        }
    }
}